### ISSUE TITLE
Road shields / markers get a wrong color based on a letter presence in a ref=*

### ISSUE THREAD

**Issue by pastk** [organicmaps/organicmaps#6170]
https://github.com/organicmaps/organicmaps/commit/43cf2383202 changed refs parsing logic so that now e.g. presence of a letter `B` in a `FFB 11` ref (Germany) makes it orange (should be white), see https://github.com/organicmaps/organicmaps/issues/4881#issuecomment-1742020033

Apparently it was done to fix some other cases (in France? see the above commit and a linked PR for details).

It changed color of `A1 (M)` from blue to yellow-on-green in UK too (but its actually seems to be more complicated there and a right blue color was used before just by coincidence, see https://github.com/organicmaps/organicmaps/issues/1244#issuecomment-1741726968).

CC @vng

**Comment by vng** [comment]
Can't say now, should check. You can modify this logic (not find, but StartsWith), but check referenced issues:
https://github.com/organicmaps/organicmaps/issues/2581
https://github.com/organicmaps/organicmaps/issues/2582

**Comment by biodranik** [comment]
@pastk is it possible to make a unit test to avoid breaking something after changes?

**Comment by pastk** [comment]
> @pastk is it possible to make a unit test to avoid breaking something after changes?

Sure! And Victor has added first tests with that change, before that we had none so it was very easy to break something as not possible to test in all countries :)
And we'll expand the test coverage over the time.

**Comment by telluroid** [comment]
> Can't say now, should check. You can modify this logic (not find, but StartsWith), but check referenced issues: #2581 #2582

StartsWith could be a problem as well because some roads seem to have the ref AIC (https://de.m.wikipedia.org/wiki/Liste_der_Kreisstra%C3%9Fen_im_Landkreis_F%C3%BCrstenfeldbruck) and there might be other examples.

**Comment by pastk** [comment]
Ah I see then Germany's `AIC` road might be colored as an `A` road, which is wrong.

One solution would be to add all possible code schemes like `AIC`, `FFB`, etc. and prefer a longer StartsWith substring. But this is tedious.

A better approach would be to use more advanced pattern matching like "`A` + any number".
An easiest would be just to use regex, but ATM road shields are parsed in the runtime. Which is wrong anyway and its better to move it into the generator, but its quite a refactoring and will require mwm version change.

**Comment by pastk** [comment]
In the interim we might add a parser class that will match a pre-defined substring in the beginning of the road only in case its followed by a digit or a whitespace or `-` (for cases like `D-106A`, `St 3611`, etc.).

**Comment by biodranik** [comment]
@pastk let's focus on the best solution without any possible mistakes or errors in each country. We're doing the best maps, right? ;-)

**Comment by sostock** [comment]
For Germany, determining the correct shield just based on the `ref` value is unfortunately not possible (in very few cases), because Kreisstraßen (district roads) in the Augsburg district are named `A <number>`. Example:
* [Kreisstraße A 1](https://www.openstreetmap.org/relation/1425334) (should have a white shield, currently blue)
* [Autobahn A1](https://www.openstreetmap.org/relation/20904) (blue shield)

Both of these roads have `ref=A 1`.

This only applies to the 33 district roads listed [here](https://de.wikipedia.org/wiki/Liste_der_Kreisstraßen_im_Landkreis_Augsburg) (A 25 doesn’t exist anymore).

Possible solutions:

* All relations with `network=BAB`, and only those, get a blue shield. However, only the actual motorways are part of the relation, motorway links (`highway=motorway_link`) are not. Some `motorway_link`s have a `ref` (that matches the `destination:ref`), so those would get a wrong (white) shield because they are not in the relation. Only the relations have `network=BAB`, the ways do not.
* Distinguish them by the `highway` value: the district roads are `secondary` or `tertiary`, the autobahns are `motorway` or `trunk`.

> In the interim we might add a parser class that will match a pre-defined substring in the beginning of the road only in case its followed by a digit or a whitespace or `-` (for cases like `D-106A`, `St 3611`, etc.).

This seems like a good interim solution to me. Only the 33 roads in the Augsburg district will get a wrong shield, all other roads in Germany should be handled correctly by this.

**Comment by biodranik** [comment]
@pastk is it possible to hard-code proper shields in the generator for these 33 roads?

**Comment by pastk** [comment]
Yes. There is no config file for shields, its all basically code. For some countries generic parsers are sufficient, while other countries (e.g. US) resort to custom parsers.
